gridstatus/miso_api.py

```diff
@@ -0,0 +1,322 @@
+        )
+
+        data_list = self._get_url(url, product=PRICING_PRODUCT, verbose=verbose)
+
+        return data_list
+
+    @support_date_range(frequency="HOUR_START", convert_to_dataframe=False)
+    def _get_lmp_real_time_hourly_ex_post(
+        self,
+        date,
+        end=None,
+        prelim_or_final: str = PRELIMINARY_STRING,
+        verbose=False,
+    ):
+        # 0-padded hour. 00 doesn't exist so add 1 to the hour
+        interval = str(date.hour + 1).zfill(2)
+        date_str = date.strftime("%Y-%m-%d")
+        version = EX_POST
+        resolution = HOURLY_RESOLUTION
+
+        url = f"{BASE_PRICING_URL}/real-time/{date_str}/lmp-{version}?interval={interval}&preliminaryFinal={prelim_or_final}&timeResolution={resolution}"  # noqa
+
+        data_list = self._get_url(url, product=PRICING_PRODUCT, verbose=verbose)
+
+        return data_list
+
+    @support_date_range(frequency="5_MIN", convert_to_dataframe=False)
+    def _get_lmp_real_time_5_min_ex_ante(self, date, end=None, verbose=False):
+        # Interval format is hh:mm at the start of the interval
+        interval = date.floor("5min").strftime("%H:%M")
+        date_str = date.strftime("%Y-%m-%d")
+        version = EX_ANTE
+
+        url = (
+            f"{BASE_PRICING_URL}/real-time/{date_str}/lmp-{version}?interval={interval}"
+        )
+
+        data_list = self._get_url(url, product=PRICING_PRODUCT, verbose=verbose)
+
+        return data_list
+
+    @support_date_range(frequency="5_MIN", convert_to_dataframe=False)
+    def _get_lmp_real_time_5_min_ex_post(
+        self,
+        date,
+        end=None,
+        prelim_or_final: str = PRELIMINARY_STRING,
+        verbose=False,
+    ):
+        # Interval format is hh:mm at the start of the interval
+        interval = date.floor("5min").strftime("%H:%M")
+        date_str = date.strftime("%Y-%m-%d")
+        version = EX_POST
+        resolution = FIVE_MINUTE_RESOLUTION
+
+        url = f"{BASE_PRICING_URL}/real-time/{date_str}/lmp-{version}?interval={interval}&preliminaryFinal={prelim_or_final}&timeResolution={resolution}"  # noqa
+
+        data_list = self._get_url(url, product=PRICING_PRODUCT, verbose=verbose)
+
+        return data_list
+
+    def _process_pricing_data(self, data_list: List[Dict], market: Markets):
+        df = pd.DataFrame(data_list)
+
+        # Split timeInterval dict into separate columns for 'start' and 'end'
+        df = pd.concat(
+            [
+                df["timeInterval"].apply(pd.Series)[["start", "end"]],
+                df.drop(columns=["timeInterval"]),
+            ],
+            axis=1,
+        )
+
+        df["Interval Start"] = pd.to_datetime(df["start"]).dt.tz_localize(
+            self.default_timezone,
+        )
+        df["Interval End"] = pd.to_datetime(df["end"]).dt.tz_localize(
+            self.default_timezone,
+        )
+
+        node_to_type_mapping = (
+            MISO()
+            ._get_node_to_type_mapping()
```

Comment:
our approach here is to query the `{today}_da_expost_lmp.csv` file for to obtain node maps. this feels unreliable for historical days when nodes may have been different. that file seems to have data going back to `20220101`. perhaps we can can pass the date of the end of the requested range (but not earlier than `20220101`) to this function to get a better file.

Reply:
I pulled the data from 2022-01-01 and 2024-12-11 and compared the mapping of `Node` to `Type`. There were no differences. (The 2024 data has more nodes but none of the mappings have changed).